brave/brave-core
# Brave search conversion promotion in omnibox

Resolves https://github.com/brave/brave-browser/issues/22616

This PR is for showing brave search promotion match in omnibox popup.
To do this, PromotionProvider adds promotion matches and BraveOmniboxResultView
renders promotion UI. There are two types of conversion. One is button type and the other is banner.
Button type match is located as a second match and Bannter type is located as a last match.
There position is re-arragned after upstreams match sorting.

First commit is `PromotionProvider`(non UI) part and second commit is `BraveOmniboxResultView` (UI) part.

Button type
<img width="1104" alt="Screen Shot 2022-06-15 at 1 17 58 PM" src="https://user-images.githubusercontent.com/6786187/173735879-aa675261-0e99-43eb-8008-fe63bd0d130a.png">
<img width="1104" alt="Screen Shot 2022-06-15 at 1 24 33 PM" src="https://user-images.githubusercontent.com/6786187/173736484-92f91a4e-f2ab-4555-bbca-be8d5e1f2a5f.png">
<img width="1104" alt="Screen Shot 2022-06-15 at 1 18 41 PM" src="https://user-images.githubusercontent.com/6786187/173735883-ae8f6be4-a4b6-4535-a610-8e2af5e89caa.png">
<img width="1103" alt="Screen Shot 2022-06-15 at 1 25 00 PM" src="https://user-images.githubusercontent.com/6786187/173736508-44f7199a-d281-42db-ba7d-318954c6a1de.png">

Banner type
<img width="1102" alt="Screen Shot 2022-06-15 at 1 20 24 PM" src="https://user-images.githubusercontent.com/6786187/173736115-101db3ca-27f5-4daa-8829-de743cf0bbbe.png">
<img width="1106" alt="Screen Shot 2022-06-15 at 1 22 01 PM" src="https://user-images.githubusercontent.com/6786187/173736311-b0d55d43-0953-4a96-8033-655ca988f61d.png">
<img width="1102" alt="Screen Shot 2022-06-15 at 1 20 07 PM" src="https://user-images.githubusercontent.com/6786187/173736137-d7b972a4-b52f-458e-8f97-78665b837cd1.png">
<img width="1097" alt="Screen Shot 2022-06-15 at 1 22 52 PM" src="https://user-images.githubusercontent.com/6786187/173736338-c82f30a5-9d54-49e2-8d0f-6a14a639a287.png">

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
Button type
1. Launch browser with `--enable-features=BraveSearchOmniboxButton` command line switch
2. Set any search engine except brave search as a default search provider
3. Type any keyword and check button type autocomplete is visible as a second item

Banner type
1. Launch browser with `--enable-features=BraveSearchOmniboxBanner` command line switch
2. Set any search engine except brave search as a default search provider
3. Type any keyword and check button type autocomplete is visible as a last item

Note:
* When first item is not keyword match from default provider, promotion match will not be added because first item is highly recommended match
* When dismissed button or x button clicked, autocomplete match list are refreshed with same input data and promotion match will not be shown anymore.
* When user types an input that was tried by promotion match, promotion match will not be shown again. Instead promotion url could be visible as a history match.
* This conversion promotion is only enabled for below 5 countries
  * US, CA, DE, FR, GB
* Only normal profile window give promotion. Private(or tor) window don't show promotion match.
